### PR TITLE
Allow a CombinedLoader as the training data in DDP

### LINKED ISSUE #11632: CombinedLoader for training data does not work in DDP
## 🐛 Bug

When using a `CombinedLoader` for the training data in conjunction with DDP, all GPUs get to see the full data.
This is pretty much the same problem as reported previously in #10373, just this time for the training data.

@tchaton FYI

### To Reproduce

https://gist.github.com/ant0nsc/ad44d277eaca209d4e387b8ba7717dcb

### Expected behavior

In the above script, CombinedLoader(..., mode="min_size") should give 6 batches. When split across 2 GPUs, each GPU should see 3 batches. However, we are seeing all 6 batches per GPU. The same problem applies when using "max_size_cycle"

### Environment
* CUDA:
        - GPU:
                - Tesla K80
                - Tesla K80
        - available:         True
        - version:           11.1
* Packages:
        - numpy:             1.19.1
        - pyTorch_debug:     False
        - pyTorch_version:   1.8.0
        - pytorch-lightning: 1.5.5
        - tqdm:              4.62.3
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - 
        - processor:         x86_64
        - python:            3.7.3
        - version:           #58~18.04.1-Ubuntu SMP Wed Jul 28 23:14:18 UTC 2021

cc @tchaton @rohitgr7 @akihironitta

### PR DESCRIPTION
**Pull request by ant0nsc**
## What does this PR do?

This PR allows to use a `CombinedLoader` as training data, and ensures that the right samplers are added when used in DDP

Fixes #11632
Fixes: #11475

### Does your PR introduce any breaking changes? If yes, please list them.

None that I'm aware of.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -1769,26 +1769,35 @@ def reset_train_dataloader(self, model: Optional["pl.LightningModule"] = None) -
...
+        else:
+            self.train_dataloader = apply_to_collection(
+                self.train_dataloader,
+                DataLoader,
+                self._data_connector._prepare_dataloader,
+                shuffle=True,
+                mode=RunningStage.TRAINING,
+            )
+            is_combined_loader = False
+        loaders = self.train_dataloader.loaders if is_combined_loader else self.train_dataloader
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:1786]:
```suggestion
        self.train_dataloader = apply_to_collection(
            self.train_dataloader,
            DataLoader,
            self._data_connector._prepare_dataloader,
            shuffle=True,
            mode=RunningStage.TRAINING,
        )
        loaders = self.train_dataloader.loaders if isinstance(self.train_dataloader, CombinedLoader) else self.train_dataloader
```

Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py:1786]: _prepare_dataloader can handle combined loaders

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1769,26 +1769,35 @@ def reset_train_dataloader(self, model: Optional["pl.LightningModule"] = None) -
...
+        apply_to_collection(loaders, DataLoader, _auto_add_worker_init_fn, rank=self.global_rank)
 
         # add collate_fn to collect metadata for fault tolerant training
         if _fault_tolerant_training():
-            apply_to_collection(self.train_dataloader, DataLoader, _add_capture_metadata_collate)
+            apply_to_collection(loaders, DataLoader, _add_capture_metadata_collate)
 
         # wrap the sequence of train loaders to a CombinedLoader object for computing the num_training_batches
-        self.train_dataloader = CombinedLoader(self.train_dataloader, self._data_connector.multiple_trainloader_mode)
+        if not is_combined_loader:
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:1799]:
```suggestion
        if not isinstance(self.train_dataloader, CombinedLoader):
```

**Review by rohitgr7 — COMMENTED** [review]
LG!

**Comment by rohitgr7** [comment]
also fixes: https://github.com/PyTorchLightning/pytorch-lightning/issues/11475?

also, would you mind updating the docs too: https://pytorch-lightning.readthedocs.io/en/latest/guides/data.html#return-multiple-dataloaders?

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (ant0nsc) [reply pytorch_lightning/trainer/trainer.py]: I tried your suggestion, but it breaks the tests, DDP is then not handled correctly for the CombinedLoader.

Comment (ant0nsc) [reply pytorch_lightning/trainer/trainer.py]: got it to work with a small change, thanks!

Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py]: ah! my bad. forgot to add that

**Comment by ant0nsc** [comment]
Can I have more reviews here please? @rohitgr7 @tchaton for example. Also, one testing pipeline timed out, I did not find a way of re-queuing it.

**Comment by rohitgr7** [comment]
> Can I have more reviews here please? @rohitgr7 @tchaton for example. Also, one testing pipeline timed out, I did not find a way of re-queuing it.

hey @ant0nsc, you can ignore the timeout. happens sometimes. will review this on Monday, although looks great. Just needs some more doc updates. Will push some commits if required.

File: tests/trainer/test_supporters.py [hunk 3]
```diff
@@ -382,7 +382,7 @@ def _assert_dataset(loader):
 
 
 @pytest.mark.parametrize("replace_sampler_ddp", [False, True])
-def test_combined_data_loader_with_max_size_cycle_and_ddp(replace_sampler_ddp, tmpdir):
+def test_combined_data_loader_with_max_size_cycle_and_ddp(replace_sampler_ddp):
     """This test makes sure distributed sampler has been properly injected in dataloaders when using CombinedLoader
     with ddp and `max_size_cycle` mode."""
     trainer = Trainer(strategy="ddp", accelerator="auto", devices=2, replace_sampler_ddp=replace_sampler_ddp)
```
Comment (tchaton) [inline tests/trainer/test_supporters.py:388]:
```suggestion
    trainer = Trainer(default_root_dir=tmpdir, strategy="ddp", accelerator="auto", devices=2, replace_sampler_ddp=replace_sampler_ddp)
```

File: tests/trainer/test_supporters.py [hunk 4]
```diff
@@ -382,7 +382,7 @@ def _assert_dataset(loader):
 
 
 @pytest.mark.parametrize("replace_sampler_ddp", [False, True])
-def test_combined_data_loader_with_max_size_cycle_and_ddp(replace_sampler_ddp, tmpdir):
+def test_combined_data_loader_with_max_size_cycle_and_ddp(replace_sampler_ddp):
```
Comment (tchaton) [inline tests/trainer/test_supporters.py:385]: Better to always use tmpdir.

**Review by tchaton — APPROVED** [review]
LGTM !

File: tests/trainer/test_supporters.py (hunk 4 above)
Comment (ant0nsc) [reply tests/trainer/test_supporters.py:385]: done, also in my test

**Review by rohitgr7 — APPROVED** [review]
added an example

File: docs/source/guides/data.rst [hunk 5]
```diff
@@ -184,6 +186,24 @@ Furthermore, Lightning also supports nested lists and dicts (or a combination).
...
+        loader_a = DataLoader()
+        loader_b = DataLoader()
+        loaders = {"a": loader_a, "b": loader_b}
+        combined_loader = CombinedLoader(loaders, mode="max_size_cycle")
+        return combined_loader
+
+
+    def training_step(self, batch, batch_idx):
+        batch_a = batch["a"]
+        batch_b = batch["b"]
```
Comment (carmocca) [inline docs/source/guides/data.rst:206]: nit: there were 2 line breaks here

```suggestion
        batch_b = batch["b"]

```

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]
